### PR TITLE
Dynamically load resolution checks

### PR DESCRIPTION
**Pull request by ludeeus**
## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (which adds functionality to the supervisor)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast supervisor tests`)
- [ ] Tests have been added to verify that the new code works.

If API endpoints of add-on configuration are added/changed:

- [ ] Documentation added/updated for [developers.home-assistant.io][docs-repository]

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[docs-repository]: https://github.com/home-assistant/developers.home-assistant

### CODE REVIEW

File: supervisor/resolution/checks/base.py [hunk 1]
```diff
@@ -17,6 +18,13 @@ def __init__(self, coresys: CoreSys) -> None:
         """Initialize the checks class."""
         self.coresys = coresys
 
+    def __init_subclass__(cls, **kwargs):
+        """Initialize subclasses."""
+        super().__init_subclass__(**kwargs)
+        check = cls(None)
+        if check.slug not in CHECK_REGISTRY:
+            CHECK_REGISTRY.append(check)
+
```
Comment (pvizeli) [inline supervisor/resolution/checks/base.py:27]: I think how Home Assistant does it and also perl/go kind of using it for dynamic loading is like https://github.com/home-assistant/supervisor/blob/main/supervisor/discovery/validate.py#L23 (Code is HA) and add an function like `def setup() -> CheckBase:` which would call to get it from. That way it need also less code

Comment (ludeeus) [reply supervisor/resolution/checks/base.py:27]: neat!

**Comment by ludeeus** [comment]
Once this is merged, we should use the same approach for fixups and evaluations as well

File: supervisor/resolution/check.py [hunk 2]
```diff
@@ -19,9 +18,8 @@ class ResolutionCheck(CoreSysAttributes):
     def __init__(self, coresys: CoreSys) -> None:
         """Initialize the checks class."""
         self.coresys = coresys
-        self._core_security = CheckCoreSecurity(coresys)
-        self._free_space = CheckFreeSpace(coresys)
-        self._addon_pwned = CheckAddonPwned(coresys)
+        self._checks: Dict[str, CheckBase] = {}
+        self.load()
```
Comment (pvizeli) [inline supervisor/resolution/check.py:22]:
```suggestion
        self._load()
```

Is not something that should call out of this class or multiple times

File: supervisor/resolution/check.py [hunk 3]
```diff
@@ -31,14 +29,23 @@ def data(self) -> Dict[str, Any]:
     @property
     def all_checks(self) -> List[CheckBase]:
         """Return all list of all checks."""
-        return [self._core_security, self._free_space, self._addon_pwned]
+        return list(self._checks.values())
+
+    def load(self):
+        """Load all checks."""
```
Comment (pvizeli) [inline supervisor/resolution/check.py:35]:
```suggestion
    def _load(self):
        """Load all checks."""
```

File: supervisor/resolution/check.py [hunk 4]
```diff
@@ -31,14 +29,23 @@ def data(self) -> Dict[str, Any]:
...
+        package = f"{__package__}.checks"
+        for module in get_valid_modules("checks"):
+            check_module = import_module(f"{package}.{module}")
+            check = check_module.setup(self.coresys)
+
+            if check.slug not in self._checks:
+                self._checks[check.slug] = check
```
Comment (pvizeli) [inline supervisor/resolution/check.py:42]:
```suggestion
            if check.slug not in self._checks:
                self._checks[check.slug] = check
```
```suggestion
            self._checks[check.slug] = check
```

**Review by pvizeli — COMMENTED** [review]
Can be merge after my suggestion 👍

File: supervisor/resolution/check.py [hunk 5]
```diff
@@ -31,14 +29,21 @@ def data(self) -> Dict[str, Any]:
...
-        return [self._core_security, self._free_space, self._addon_pwned]
+        return list(self._checks.values())
+
+    def _load(self):
+        """Load all checks."""
+        package = f"{__package__}.checks"
+        for module in get_valid_modules("checks"):
+            check_module = import_module(f"{package}.{module}")
+            check = check_module.setup(self.coresys)
+            self._checks[check.slug] = check
```
Comment (ludeeus) [inline supervisor/resolution/check.py:40]:
```suggestion
            self._checks[check.slug] = check_module.setup(self.coresys)
```
No need to define `check` here

**Pull request merged**